### PR TITLE
Add gasPrice calculation in runtime tx detail view

### PR DESCRIPTION
**Pull request by lubej**
## Description

Quote from API docs: "For EVM transactions this is calculated as `charged_fee` = `gas_price * gas_used`, where `gas_price = fee / gas_limit`, for compatibility with Ethereum."

## Resources 

image

### CODE REVIEW

**Comment by lubej** [comment]
@donouwens could you take a quick glance on how the Gas price is displayed. Particularly the `(0.0000001 ROSE)` part.

File: src/app/utils/number-utils.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import BigNumber from 'bignumber.js'
+
+const GWEI_DENOMINATION = BigNumber(10).pow(9)
+
+export const getGasPrice = ({ fee = '', gasUsed = '' }): string | null => {
+  if (!fee || !gasUsed) {
+    return null
+  }
+
+  return BigNumber(fee).div(BigNumber(gasUsed)).toLocaleString()
```
Comment (lubej) [inline src/app/utils/number-utils.ts:10]: Can `toLocaleString` brake `getPreciseNumberFormat`?

File: src/app/utils/number-utils.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import BigNumber from 'bignumber.js'
+
+const GWEI_DENOMINATION = BigNumber(10).pow(9)
+
+export const getGasPrice = ({ fee = '', gasUsed = '' }): string | null => {
+  if (!fee || !gasUsed) {
```
Comment (lukaw3d) [inline src/app/utils/number-utils.ts:6]: I'd add `gasUsed === '0'`

File: src/app/utils/number-utils.ts (hunk 1 above)
Comment (lukaw3d) [reply src/app/utils/number-utils.ts:10]: Maybe. It's odd that it even works
- `toLocaleString` comes from Object in BigNumber's prototype chain
- Object.toLocaleString natively calls `this.toString`
- BigNumber.toString sometimes returns exponential format (1e+28, 1.234e-35).

toFixed is the safest option
https://github.com/oasisprotocol/explorer/blob/ed691bb3a65963b1c8345dd279454cab56e70b72/src/app/utils/helpers.ts#L113-L119

File: src/app/pages/RuntimeTransactionDetailPage/index.tsx [hunk 3]
```diff
@@ -311,6 +314,27 @@ export const RuntimeTransactionDetailView: FC<{
             })}
           </dd>
 
+          {gasPrice && (
+            <>
+              <dt>{t('common.gasPrice')}</dt>
+              <dd>
+                {t('common.valueInToken', {
+                  ...getPreciseNumberFormat(formatGwei(gasPrice)),
+                  ticker: 'Gwei',
```
[the remark this answers is not in the document]
Comment (lukaw3d) [reply src/app/pages/RuntimeTransactionDetailPage/index.tsx:323]: Not sure if we have some good established format for printing gas price. Ask internally

https://docs.oasis.io/general/manage-tokens/cli/addressbook/#add
oasis CLI prints "(gas price: 0.0000001 ROSE per gas unit)"

File: src/app/utils/number-utils.ts (hunk 1 above)
Comment (lubej) [reply src/app/utils/number-utils.ts:10]: Yes, `toFixed` looks good. Thanks 👍 

It's true, that `toString` returns exponent format, but looks like `Intl.NumberFormat` correctly parsed it.

File: src/app/pages/RuntimeTransactionDetailPage/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (lubej) [reply src/app/pages/RuntimeTransactionDetailPage/index.tsx:323]: Yes, it does not work for all ParaTimes. But Gwei is something I would expect on Sapphire chain, but "breaks" context on Cipher. I would definitely skip `per gas unit` wording, as that is a given.

Can we maybe display Gwei on ParaTimes that have `type === RuntimeTypes.Evm` in config and `18 decimal spaces` denomination?

File: src/app/utils/number-utils.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+export const getGasPrice = ({ fee = '', gasUsed = '' }): string | null => {
+  if (!fee || !gasUsed || gasUsed === '0') {
+    return null
+  }
+
+  return BigNumber(fee).div(BigNumber(gasUsed)).toFixed()
+}
+
+export const fromBaseUnitsToNano = (wei: string): string => fromBaseUnits(wei, -9)
```
Comment (lukaw3d) [inline src/app/utils/number-utils.ts:19]: This is not taking in base units

Comment (lubej) [reply src/app/utils/number-utils.ts:19]: True. But what would wei unit be called then?

Comment (lubej) [reply src/app/utils/number-utils.ts:19]: Thinking about it now, you probably meant the function parameter, not the naming. Nevermind.

Comment (lukaw3d) [reply src/app/utils/number-utils.ts]: > `export const fromBaseUnitsToNano = (valueInBaseUnits: string): string => fromBaseUnits(valueInBaseUnits, -9)`

Both. This takes in values like "0.0000001", clearly not in base units

File: src/app/pages/RuntimeTransactionDetailPage/index.tsx [hunk 5]
```diff
@@ -311,6 +314,18 @@ export const RuntimeTransactionDetailView: FC<{
             })}
           </dd>
 
+          {gasPrice && (
+            <>
+              <dt>{t('common.gasPrice')}</dt>
+              <dd>
+                {t('common.valueInToken', {
+                  ...getPreciseNumberFormat(fromBaseUnitsToNano(gasPrice)),
+                  ticker: `n${tickerName}`,
```
Comment (lukaw3d) [inline src/app/pages/RuntimeTransactionDetailPage/index.tsx:323]: Add TODO based on Jernej's comments

File: src/app/utils/number-utils.ts (hunk 4 above)
Comment (lubej) [reply src/app/utils/number-utils.ts]: Ah right, for some reason I though on Sapphire 1 ROSE = baseUnit 🤦 

Well should I just name it `unit` then? Or maybe `exaBaseUnit` - but that one is also not correct, as it depends on the ParaTime?

Comment (lubej) [reply src/app/utils/number-utils.ts]: Did generalize, I hope it's ok now.

Comment (lukaw3d) [reply src/app/utils/number-utils.ts]: :D 
(maybe toBaseUnits)

**Review by lukaw3d — APPROVED** [review]

File: src/app/utils/number-utils.ts (hunk 4 above)
Comment (lubej) [reply src/app/utils/number-utils.ts]: `toBaseUnit` would imply that "conversion" is done by 10^18, not by 10^9. Either way, I am keeping it as is.

**Pull request merged**